### DISCUSSION TITLE
In my recent Airflow exploration, I'm trying to dump logs from Airflow to external systems, possibly without using FluentD, the target system is Splunk in my case.

### DISCUSSION THREAD

**Discussion by skyboi1233** [apache/airflow#24648] [Q&A]
In my recent Airflow exploration, I'm trying to dump logs from Airflow to external systems, possibly without using FluentD, the target system is Splunk in my case.
I was wondering if there is a way to add metainformation like DAG_ID, TASK_ID or RUN_ID to my task's logs in order to provide them with an internal index.
I can see a AIRFLOW__LOGGING__LOG_FORMAT env.var exists, btw it seems like you cannot use Airflow internal variable in this scope but only the Python logger ones.
Can anyone suggest a way to achieve that?

_Originally posted by @MattiaGallegati in https://github.com/apache/airflow/discussions/24633_

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
This ois something that we are going to have in the future once we integrate with Open -Telemetry and traces. https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-49+OpenTelemetry+Support+for+Apache+Airflow  - the Airflow Improvement Proposal has just been merged but it will take quite a while to implement. You'd have to do the job that is planned there to achieve it.

**Reply by MattiaGallegati** [reply]
Thank you so much! So you basically confirm that right now there is no "native" way to append context variable such as Run_Id/Task_Id ecc.. at least to task log using some env like the AIRFLOW__LOGGING__LOG_FORMAT right?
The only "workaround" for that at the moment is to append these ids building them from dir path of the logs using Logstash or FluentD correct?

**Reply by potiuk** [reply]
I think i depends. Now when I am reading your question I am not sure any more if I properly understand your question. In either case the answer is (as usual "it depends") . 

If you are interested in just "pure task" logs (and not other logs connected wiht running the tasks/dags/run_id - which might be generated in various parts of Airflow) - this is fully is possible and [documented](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/logging-tasks.html) and there are [a number of log handlers](https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/logging.html) implemented in Airflow that can tell you "this log comes from this particular run_id of this task_id from this dag_id". - basically if you are only interested in the same logs that you see in UI of Airlfow when you look at task-details this is fully covered. Custom log handlers can be easily added using those we have in Airflow as examples.

However if you are interested in seeing logs from other part of the system (say scheduler scheduling particular run_id or particular task_id of particular dag_id) - this is somethign that we are aiming for at the "traces" of Open-Telemetry and there is no "easy and documented" way - however you can certainly build custom handlers and filters and build it on your own if you are interested. We are using standard Python logger framework, you can customise which handlers are installed and you can add your own handlers for any kind of log being written, so if you **really want** to do it - you can. This is "native" in the sense there is nothing special - just installing custom logging handlers if you want them "custom-handled" (which is generally approach that any python app follows).  But it's not "easy" and "configurale"  - you have to invest into it by finding and filtering the right logs and custom handling them.

In the sense - coming Open Telemetry support will be both "native" and "easy" (once completed) - because we will use OpenTelemetry traces API - Airlfow will be producing right traces. If your system (say splunk) will be able to consume them, it will be just the matter of configuration.

So you are not "blocked" by anything there - it's just much more effort from your side now than it will be some time in the future.

**Reply by MattiaGallegati** [reply]
Thank you for the detailed response.
I had a look at the manual before posting the question. 
The thing is that I'm not able to understand clearly the way to do that. Just suppose I want to have my task logs saved to filesystem but with a unique ID made from RUN_ID-DAG_ID-TASK_ID ecc.. I'm still wondering what is the "simple way" to achieve that.
I can see the AIRFLOW__LOGGING__LOG_FORMAT but it seems that it can't take these kind of values.
I can also see different implementation of loggers from providers but in my case I will have no ELK or OSS or S3 to enable these providers.
Do I have to implement my own log_handler to attach this information?

**Reply by potiuk** [reply]
yes

**Reply by brki** [reply]
A bit late, but perhaps `airflow.utils.log.task_handler_with_custom_formatter.TaskHandlerWithCustomFormatter` could be helpful.
